src/model/Game.ts

```diff
@@ -0,0 +1,409 @@
+			firstEdge = this.getRightEdge(x, y)
+			secondEdge = this.getRightEdge(x, y + 1)
+		}
+
+		// trying to place wall
+		this.gridEdges.delete(firstEdge)
+		this.gridEdges.delete(secondEdge)
+
+		let winConditionsIsAccessible = true
+
+		for (const player of this.players) {
+			winConditionsIsAccessible =
+				winConditionsIsAccessible && this.checkWinCondition(player.position, player.goal)
+		}
+
+		if (!winConditionsIsAccessible) {
+			this.gridEdges.add(firstEdge)
+			this.gridEdges.add(secondEdge)
+			console.log("Wall blocks someone`s path")
+			//throw new Error("Wall blocks someone`s path")
+		}
+
+		const currentPlayer = this.getCurrentPlayer()
+		currentPlayer.walls-- // reducing player wall counter
+		this.wallsAvailable.delete(move.position)
+
+		let wall1, wall2, wall3
+
+		if (orientation === "h") {
+			wall1 = this.stringToWall((x + 1), y, "h")
+			wall2 = this.stringToWall((x - 1), y, "h")
+			wall3 = this.stringToWall(x, y, "v")
+		} else {
+			wall1 = this.stringToWall(x, (y + 1), "v")
+			wall2 = this.stringToWall(x, (y - 1), "v")
+			wall3 = this.stringToWall(x, y, "h")
+		}
+
+		if (this.wallsAvailable.has(wall1)) {
+			this.wallsAvailable.delete(wall1)
+			move.removedWalls.push(wall1)
+		}
+		if (this.wallsAvailable.has(wall2)) {
+			this.wallsAvailable.delete(wall2)
+			move.removedWalls.push(wall2)
+		}
+		if (this.wallsAvailable.has(wall3)) {
+			this.wallsAvailable.delete(wall3)
+			move.removedWalls.push(wall3)
+		}
+	}
+
+	undoLastMove(): void {
+		const lastMove = this.moveHistory.pop()
+		if (lastMove !== undefined) {
+			this.playerToMakeMove = this.playerToMakeMove === 0 ? 1 : 0
+			const currentPlayer = this.getCurrentPlayer()
+
+			if (isPlayerMove(lastMove)) {
+				if (lastMove.previousPosition != null) { // typescript go brr
+					currentPlayer.position.x = lastMove.previousPosition.x
+					currentPlayer.position.y = lastMove.previousPosition.y
+				}
+			} else {
+				const { x, y, orientation } = this.wallToString(lastMove.position)
+				// restore available moves
+				lastMove.removedWalls.forEach(e => this.wallsAvailable.add(e))
+
+				let firstEdge, secondEdge
+				if (orientation === "h") {
+					firstEdge = this.getBottomEdge(x, y)
+					secondEdge = this.getBottomEdge(x + 1, y)
+				} else {
+					firstEdge = this.getRightEdge(x, y)
+					secondEdge = this.getRightEdge(x, y + 1)
+				}
+
+				// restoring edges
+				this.gridEdges.add(firstEdge)
+				this.gridEdges.add(secondEdge)
+			}
+		}
+	}
+
+	getCurrentPlayer(): Player{
+		return this.players[this.playerToMakeMove]
+	}
+
+	getNextPlayer(): Player {
+		return this.playerToMakeMove === 0 ? this.players[1] : this.players[0]
+	}
+
+	wallToString(wallPosition: string): {x: number, y: number, orientation: string} {
+		const decodedPosition = wallPosition.split("")
+		return {x: parseInt(decodedPosition[0]), y: parseInt(decodedPosition[1]), orientation: decodedPosition[2]}
+	}
+
+	stringToWall(x: number, y: number, orientation: string): string {
```

Comment:
I would suggest to move `{x: number, y: number, orientation: string}` into the different Type so you don't have to duplicate the code